データベース設計

deriya-a opened the pull request:
# What
・README.mdにデータベース設計の記述
・furima.dioにER図の描画

ER図の画像リンク
https://gyazo.com/dfa81acf94f81554202ee3e6ee3cb89d

# Why
データベースを設計するため

deriya-a committed:
Update README.md

deriya-a committed:
ER図作成

拡張子”.dio”でFURIMA-39154直下に作成

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| first_name_kana    | string     | null: false                   |
+| birth_year         | enum       | null: false                   |
+| birth_month        | enum       | null: false                   |
+| birth_date         | enum       | null: false                   |
```
```
| birth_year         | enum       | null: false                   |
| birth_month        | enum       | null: false                   |
| birth_date         | enum       | null: false                   |
```
誕生日のカラムは「**date型**」を用いることで年、月、日を1つにして保存できます。

現状は3つカラムを設けていただいておりますが、
1つのカラムに変更し、date型で定義しましょう！

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| condition          | enum       | null: false                    |
+| delivery_fee       | enum       | null: false                    |
+| region             | enum       | null: false                    |
+| delivery_within    | enum       | null: false                    |
```
**商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリー**のカラムは
active_hashで実装の予定のため、カラム名には`「_id」`をつけ、`integer`型にしましょう！

詳しくは「`最終課題補足カリキュラム 〜カテゴリーの選択を工夫しよう〜`」のカリキュラムを参考にしていただけると幸いです。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| delivery_fee       | enum       | null: false                    |
+| region             | enum       | null: false                    |
+| delivery_within    | enum       | null: false                    |
+| price              | string     | null: false                    |
```
```
| price              | string     | null: false                    |
```
priceカラムは**数値のみ**を保存するため、**integer型**で設計しましょう！

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
 
-* Services (job queues, cache servers, search engines, etc.)
+- belongs_to :users
+- has_one :records
```
アソシエーションにおいて、`belongs_to`と`has_one`を使用する場合は
**単語の末尾を単数形**にしましょう！

理由：
`belongs_to`は1対多を示しており、
`has_one`は1対1を示すため。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+
+- belongs_to :users
+- belongs_to :items
+- has_one :deliveries
```
52〜54行目のカラムについても、単語の末尾を単数形にしましょう！

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| Column             | Type       | Options                        |
+|--------------------|------------|--------------------------------|
+| post_code          | string     | null: false                    |
+| address1           | string     | null: false                    |
```
```
| address1           | string     | null: false                    |
```
こちらの都道府県カラムは、
active_hashのデータとして使い回すことができますので、
itemsテーブルの都道府県カラムと同じものを使用しましょう。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+
+### Association
+
+- belongs_to :records
```
`- belongs_to :records`

上記に関しても、単語の末尾を単数形に修正しましょう！

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| address1           | string     | null: false                    |
+| address2           | string     | null: false                    |
+| address3           | text       | null: false                    |
+| address4           | text       | null: true                     |
```
```
| address4           | text       | null: true                     |
```
上記は建物名のカラムでお間違いないでしょうか？

その場合、`null: true`の記述は削除（空欄）にしていただいて大丈夫です。

理由：空でも登録できるようにする場合は、そもそものOptionsが不要であるため。

deriya-a replied:
記述を変更しました。

ER図を変更しました。
https://gyazo.com/f1cc623b28a50aeab55f09d4675c44f7

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
とても良い設計になっていて、大変素晴らしいです。

ER図を確認の上、コメントさせていただきましたので
ご確認よろしくお願いいたします。

`また、次回ご提出の際にはER図も修正が必要となります。`
`お手数ではありますが、修正後のER図を再度こちらのプルリクエストに添付していただくようお願いいたします。`

deriya-a committed:
Update README.md

・priceカラム名をinteger型へ更新
・belongs_toとhas_oneに続く単語の末尾を単数形へ更新
・null: trueの記述は削除
・誕生日のカラムをdate型へ更新
・itemsの都道府県カラムをactive_hashを前提に更新
・商品の状態, 配送料の負担, 発送元の地域などのカラムをactive_hash前提で更新

deriya-a committed:
ER図更新

README.mdの更新に準じて変更

ushi10 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| first_name_kanji   | string     | null: false                   |
+| family_name_kana   | string     | null: false                   |
+| first_name_kana    | string     | null: false                   |
+| birthday           | data       | null: false                   |
```
```
| birthday           | data       | null: false                   |
```
細かい部分ではありますが、上記型は正しくは`date`でございます。
型の種類を修正をしましょう。

deriya-a replied:
単純なミス、失礼いたしました。ありがとうございます。

ushi10 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+
+| Column             | Type       | Options                        |
+|--------------------|------------|--------------------------------|
+| user_id            | references | null: false, foreign_key: true |
+| item_id            | references | null: false, foreign_key: true |
```
```
| user_id            | references | null: false, foreign_key: true |
| item_id            | references | null: false, foreign_key: true |
```
上記外部キーカラム名末尾の`_id`は削除しましょう。
references型を用いる場合、カラム生成時に`_id`が自動付与されるためです。

deriya-a replied:
記述を修正しました。

ushi10 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| Column             | Type       | Options                        |
+|--------------------|------------|--------------------------------|
+| post_code          | string     | null: false                    |
+| address1_id        | integer    | null: false                    |
```
```
| address1_id        | integer    | null: false                    |
```
https://github.com/deriya-a/furima--39154/pull/1#discussion_r1195014456
上記コメント内容と重複しますが、商品テーブルの都道府県カラムと同名にしましょう。
ActiveHashを使い回すことにより作業工数削減につながるためです。

deriya-a replied:
記述を修正しました。

ushi10 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| post_code          | string     | null: false                    |
+| address1_id        | integer    | null: false                    |
+| address2           | string     | null: false                    |
+| address3           | text       | null: false                    |
+| address4           | text       |                                |
```
```
| address3           | text       | null: false                    |
| address4           | text       |                                |
```
上記カラムの型はstring型にしましょう。
それぞれ長文を保存するカラムではないためです。

deriya-a replied:
記述を修正しました。

修正後のER図です。
https://gyazo.com/0a67cb8cf48bef29fe0c6588df5702d9

ushi10 reviewed (commented):
レビュー依頼ありがとうございます。修正点確認いたしました。
追加でコメントいたしましたので、ご確認ください。
修正後、コミットプッシュを確認いただき再度レビュー依頼のご提出をお願いいたします。
また、修正後最新のER図のキャプチャ画像も併せてご提出ください。

deriya-a committed:
Update README＆ER図

ushi10 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+|--------------------|------------|--------------------------------|
+| name               | string     | null: false                    |
+| detail             | text       | null: false                    |
+| category           | integer    | null: false                    |
+| condition          | integer    | null: false                    |
+| delivery_fee       | integer    | null: false                    |
+| region             | integer    | null: false                    |
+| delivery_within    | integer    | null: false                    |
```
29〜33行目は変更せずとも問題ないため前回の記述に戻しましょう。
上記カラムはActiveHashを用いて実装をするためカラム名末尾に`_id`が必要です。

deriya-a replied:
29〜33行目のカラム名に_idを追加しました。

ushi10 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| Column             | Type       | Options                        |
+|--------------------|------------|--------------------------------|
+| post_code          | string     | null: false                    |
+| region             | integer    | null: false                    |
```
```
| region             | integer    | null: false                    |
```
上記もActiveHashを用いて実装をするため、カラム名末尾に`_id`を付与しましょう。

deriya-a replied:
カラム名regionに_idを追加しました。

上記変更にともない、ER図を更新しました。
https://gyazo.com/233f37e1c7f39fb2b1b51c8e99c4b16b

ushi10 reviewed (commented):
レビュー依頼ありがとうございます。修正点確認いたしました。
追加でコメントいたしましたので、ご確認ください。
修正後、コミットプッシュを確認いただき再度レビュー依頼のご提出をお願いいたします。
また、修正後最新のER図のキャプチャ画像も併せてご提出ください。

deriya-a committed:
Update README&ER図

Activehuchのカラム名に_idを追記

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
